app/src/kscan_sideband_behaviors.c

```diff
@@ -0,0 +1,167 @@
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+struct ksbb_entry {
+    uint8_t row;
+    uint8_t column;
+    struct zmk_behavior_binding binding;
+};
+
+struct ksbb_config {
+    const struct device *kscan;
+    struct ksbb_entry *entries;
+    size_t entries_len;
+};
+
+struct ksbb_data {
+    kscan_callback_t callback;
+    bool enabled;
+};
+
+#define GET_KSBB_DEV(n) DEVICE_DT_GET(DT_DRV_INST(n)),
+
+// The kscan callback has no context with it, so we keep a static array of all possible
+// KSBBs to check when a kscan callback from the "wrapped" inner kscan fires.
+static const struct device *ksbbs[DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)] = {
+    DT_INST_FOREACH_STATUS_OKAY(GET_KSBB_DEV)};
+
+void ksbb_inner_kscan_callback(const struct device *dev, uint32_t row, uint32_t column,
+                               bool pressed) {
+    for (int i = 0; i < ARRAY_SIZE(ksbbs); i++) {
+        const struct device *ksbb = ksbbs[i];
+        const struct ksbb_config *cfg = ksbb->config;
+        struct ksbb_data *data = ksbb->data;
+
+        if (cfg->kscan != dev) {
+            continue;
+        }
+
+        for (int e = 0; e < cfg->entries_len; e++) {
+            struct ksbb_entry *entry = &cfg->entries[e];
+            if (entry->row == row && entry->column == column) {
+                struct zmk_behavior_binding_event event = {.position = INT32_MAX,
+                                                           .timestamp = k_uptime_get()};
+
+                if (pressed) {
+                    behavior_keymap_binding_pressed(&entry->binding, event);
+                } else {
+                    behavior_keymap_binding_released(&entry->binding, event);
+                }
+                return;
+            }
+        }
+
+        if (data->enabled && data->callback) {
+            data->callback(ksbb, row, column, pressed);
+        }
+    }
+}
+
+static int ksbb_configure(const struct device *dev, kscan_callback_t callback) {
+    const struct ksbb_config *cfg = dev->config;
+    struct ksbb_data *data = dev->data;
+
+    if (!callback) {
+        return -EINVAL;
+    }
+
+    data->callback = callback;
+
+#if IS_ENABLED(CONFIG_PM_DEVICE)
+    if (pm_device_wakeup_is_enabled(dev) && pm_device_wakeup_is_capable(cfg->kscan)) {
+        pm_device_wakeup_enable(cfg->kscan, true);
+    }
+#endif // IS_ENABLED(CONFIG_PM_DEVICE)
+
+    return 0;
+}
+
+static int ksbb_enable(const struct device *dev) {
+    struct ksbb_data *data = dev->data;
+    data->enabled = true;
+
+    return 0;
+}
+
+static int ksbb_disable(const struct device *dev) {
+    struct ksbb_data *data = dev->data;
+    data->enabled = false;
+
+    return 0;
+}
+
+static int ksbb_init(const struct device *dev) {
+    const struct ksbb_config *config = dev->config;
+
+    if (!device_is_ready(config->kscan)) {
+        LOG_ERR("kscan %s is not ready", config->kscan->name);
+        return -ENODEV;
+    }
+
+    kscan_config(config->kscan, &ksbb_inner_kscan_callback);
```

Comment:
Is it okay to call a kscan device's config function before it is initialized? If not, is there anything here that ensures that the kscan we're attaching to is initialized first? (I'm not seeing anything, but I don't know if Zephyr is automatically determining device's dependencies or anything.)

Reply:
If this is a problem, then maybe you could call `kscan_config()` from `ksbb_configure()` instead of `ksbb_init()`?

Reply:
> Is it okay to call a kscan device's config function before it is initialized? If not, is there anything here that ensures that the kscan we're attaching to is initialized first? (I'm not seeing anything, but I don't know if Zephyr is automatically determining device's dependencies or anything.)

The call to `device_is_ready` just above this will ensure that we're making the call into an already initialized device. Under the hood that calls `z_device_is_ready` which does:

```
bool z_device_is_ready(const struct device *dev)
{
	/*
	 * if an invalid device pointer is passed as argument, this call
	 * reports the `device` as not ready for usage.
	 */
	if (dev == NULL) {
		return false;
	}

	return dev->state->initialized && (dev->state->init_res == 0U);
}
```

we do this in the *init* because we want to catch events even if we're not configured ourselves, such as when wrapping a direct GPIO for a dedicated pin that's not used in the matrix at all.

Reply:
Oh, and right now this works because this driver is in the `APPLICATION` init level, but once we move to Zephyr 3.5 we'll probably instead need to have a custom init priority lower than the `CONFIG_KSCAN_INIT_PRIORITY`.

Reply:
Added a new dedicated init priority for this instead of using the `APPLICATION` init level, so we don't have to fix this later.